src/datanode/src/heartbeat/handler.rs

```diff
@@ -0,0 +1,176 @@
+    fn handle(
+        &self,
+        ctx: &mut HeartbeatResponseHandlerContext,
+        resp: &HeartbeatResponse,
+    ) -> Result<()>;
+}
+
+pub trait HeartbeatResponseHeadlerExecutor: Send + Sync {
+    fn handle(&self, ctx: HeartbeatResponseHandlerContext, resp: HeartbeatResponse) -> Result<()>;
+}
+
+pub struct NaiveHandlerExecutor {
+    handler: Arc<dyn HeartbeatResponseHeadler>,
+}
+
+impl NaiveHandlerExecutor {
+    pub fn new(handler: Arc<dyn HeartbeatResponseHeadler>) -> Self {
+        Self { handler }
+    }
+}
+
+impl HeartbeatResponseHeadlerExecutor for NaiveHandlerExecutor {
+    fn handle(
+        &self,
+        mut ctx: HeartbeatResponseHandlerContext,
+        resp: HeartbeatResponse,
+    ) -> Result<()> {
+        self.handler.handle(&mut ctx, &resp)
+    }
+}
+
+#[derive(Clone)]
+pub struct InstructionHandler {
+    catalog_manager: CatalogManagerRef,
+    table_engine_manager: TableEngineManagerRef,
+}
+
+impl InstructionHandler {
+    pub fn new(
+        catalog_manager: CatalogManagerRef,
+        table_engine_manager: TableEngineManagerRef,
+    ) -> Self {
+        Self {
+            catalog_manager,
+            table_engine_manager,
+        }
+    }
+
+    pub fn execute(
+        &self,
+        ctx: &mut HeartbeatResponseHandlerContext,
+        (meta, instruction): IncomingMessage,
+    ) -> Result<()> {
+        match instruction {
+            Instruction::OpenTable {
+                catalog,
+                schema,
+                table,
+                table_id,
+                engine,
+                ..
+            } => self.handle_open_table(
+                ctx,
+                meta,
+                engine,
+                OpenTableRequest {
+                    catalog_name: catalog,
+                    schema_name: schema,
+                    table_name: table,
+                    table_id,
+                },
+            ),
+            Instruction::CloseTable {
+                catalog,
+                schema,
+                table,
+                table_id,
+                engine,
+                ..
+            } => self.handle_close_table(
+                ctx,
+                meta,
+                engine,
+                CloseTableRequest {
+                    catalog_name: catalog,
+                    schema_name: schema,
+                    table_name: table,
+                    table_id,
+                },
+            ),
+        }
+        Ok(())
+    }
+}
+
+impl HeartbeatResponseHeadler for InstructionHandler {
+    fn handle(
+        &self,
+        ctx: &mut HeartbeatResponseHandlerContext,
+        resp: &HeartbeatResponse,
+    ) -> Result<()> {
+        let executable_messages = resp
+            .mailbox_messages
+            .iter()
+            .filter_map(|m| {
+                m.payload
+                    .as_ref()
+                    .map(|_| mailbox_message_to_incoming_message(m.clone()))
+            })
```

Comment:
Seems `filter_map` and `payload`'s `map` is not necessary?

```suggestion
            .map(|m| { mailbox_message_to_incoming_message(m.clone()) })
```

Reply:
`mailbox_message_to_incoming_message` will raise a error if the payload is none 🥲